File: db/behavior.go
package db

import (
	"encoding/json"
	"log"
	"math/rand"
	"os"
	"strings"
	"sync"
)

// BehaviorData 從 npc_behaviors.json 載入的全部 NPC 行為資料。
type BehaviorData struct {
	TimePeriods map[string][2]int          `json:"time_periods"`
	Roles       map[string]RoleBehavior    `json:"roles"`
}

// RoleBehavior 單一職稱的行為定義。
type RoleBehavior struct {
	Idle           map[string][]string `json:"idle"`
	EnterReactions []string            `json:"enter_reactions"`
	ShiftArrive    string              `json:"shift_arrive"`
	ShiftLeave     string              `json:"shift_leave"`
	WanderRooms    []string            `json:"wander_rooms"`
	WanderLeave    string              `json:"wander_leave"`
	WanderArrive   string              `json:"wander_arrive"`
}

var (
	behaviorOnce sync.Once
	behaviorCache *BehaviorData
)

// LoadBehaviors 讀取並快取 npc_behaviors.json；首次呼叫後即從快取返回。
func LoadBehaviors(path string) *BehaviorData {
	behaviorOnce.Do(func() {
		data, err := os.ReadFile(path)
		if err != nil {
			log.Printf("[behavior] load %s failed: %v", path, err)
			behaviorCache = &BehaviorData{
				Roles: make(map[string]RoleBehavior),
			}
			return
		}
		var bd BehaviorData
		if err := json.Unmarshal(data, &bd); err != nil {
			log.Printf("[behavior] parse %s failed: %v", path, err)
			behaviorCache = &BehaviorData{
				Roles: make(map[string]RoleBehavior),
			}
			return
		}
		behaviorCache = &bd
		log.Printf("[behavior] loaded %d roles from %s", len(bd.Roles), path)
	})
	return behaviorCache
}

// GetBehaviors 取得已快取的 BehaviorData（須先呼叫 LoadBehaviors）。
func GetBehaviors() *BehaviorData {
	if behaviorCache == nil {
		return &BehaviorData{Roles: make(map[string]RoleBehavior)}
	}
	return behaviorCache
}

// GetTimePeriod 將遊戲小時（0-23）映射為 morning/noon/evening/night。
func GetTimePeriod(gameHour int) string {
	bd := GetBehaviors()
	for name, bounds := range bd.TimePeriods {
		start, end := bounds[0], bounds[1]
		if start <= end {
			if gameHour >= start && gameHour < end {
				return name
			}
		} else {
			if gameHour >= start || gameHour < end {
				return name
			}
		}
	}
	return "night"
}

func replacePlaceholders(text, npcName string, extra map[string]string) string {
	text = strings.ReplaceAll(text, "{name}", npcName)
	for k, v := range extra {
		text = strings.ReplaceAll(text, "{"+k+"}", v)
	}
	return text
}

func pickRandom(list []string) string {
	if len(list) == 0 {
		return ""
	}
	return list[rand.Intn(len(list))]
}

// PickIdleEmote 隨機選一條指定職稱、時段的閒置動作，替換 {name}。
func PickIdleEmote(title, period, npcName string) string {
	bd := GetBehaviors()
	role, ok := bd.Roles[title]
	if !ok {
		return ""
	}
	emotes := role.Idle[period]
	if len(emotes) == 0 {
		return ""
	}
	return replacePlaceholders(pickRandom(emotes), npcName, nil)
}

// PickEnterReaction 隨機選一條進房反應，替換 {name}。
func PickEnterReaction(title, npcName string) string {
	bd := GetBehaviors()
	role, ok := bd.Roles[title]
	if !ok {
		return ""
	}
	return replacePlaceholders(pickRandom(role.EnterReactions), npcName, nil)
}

// GetShiftFlavor 取得換班敘事文本（arriving=true 為上班，false 為下班）。
func GetShiftFlavor(title, npcName string, arriving bool) string {
	bd := GetBehaviors()
	role, ok := bd.Roles[title]
	if !ok {
		return ""
	}
	if arriving {
		return replacePlaceholders(role.ShiftArrive, npcName, nil)
	}
	return replacePlaceholders(role.ShiftLeave, npcName, nil)
}

// GetWanderRooms 取得指定職稱的巡邏房間列表。
func GetWanderRooms(title string) []string {
	bd := GetBehaviors()
	role, ok := bd.Roles[title]
	if !ok {
		return nil
	}
	return role.WanderRooms
}

// GetWanderFlavor 取得巡邏敘事（leaving=true 為離開當前房間，false 為到達新房間）。
func GetWanderFlavor(title, npcName string, roomName string, leaving bool) string {
	bd := GetBehaviors()
	role, ok := bd.Roles[title]
	if !ok {
		return ""
	}
	extra := map[string]string{}
	if leaving {
		extra["dest"] = roomName
		return replacePlaceholders(role.WanderLeave, npcName, extra)
	}
	extra["from"] = roomName
	return replacePlaceholders(role.WanderArrive, npcName, extra)
}

File: src/npc/behavior.rs
//! NPC 行為文案快取（對齊既有 `npc/behavior`）。

use std::collections::HashMap;
use std::path::Path;
use std::sync::{OnceLock, RwLock};

use rand::Rng;
use serde::Deserialize;

#[derive(Debug, Deserialize, Default, Clone)]
struct RoleMovementJson {
    #[serde(default)]
    speed: i32,
}

#[derive(Debug, Deserialize, Default, Clone)]
struct RoleBehaviorJson {
    #[serde(default)]
    enter_reactions: Vec<String>,
    #[serde(default)]
    shift_arrive: String,
    #[serde(default)]
    shift_leave: String,
    #[serde(default)]
    movement: Option<RoleMovementJson>,
    #[serde(default)]
    idle: HashMap<String, Vec<String>>,
    #[serde(default)]
    wander_rooms: Vec<String>,
    #[serde(default)]
    wander_leave: String,
    #[serde(default)]
    wander_arrive: String,
}

#[derive(Debug, Deserialize, Default)]
struct BehaviorsFile {
    #[serde(default)]
    time_periods: HashMap<String, [i32; 2]>,
    #[serde(default)]
    roles: HashMap<String, RoleBehaviorJson>,
}

static BEHAVIORS: OnceLock<RwLock<Option<BehaviorsFile>>> = OnceLock::new();

fn cache() -> &'static RwLock<Option<BehaviorsFile>> {
    BEHAVIORS.get_or_init(|| RwLock::new(None))
}

/// 載入 `npc_behaviors.json`；可重複呼叫覆寫快取。
pub fn try_load_npc_behaviors(path: &Path) -> anyhow::Result<()> {
    let raw = std::fs::read_to_string(path)
        .or_else(|_| std::fs::read_to_string(Path::new("..").join(path)))?;
    let f: BehaviorsFile = serde_json::from_str(&raw)?;
    let n = f.roles.len();
    *cache().write().expect("behaviors poisoned") = Some(f);
    tracing::info!(target: "npc_behavior", "loaded {n} roles from {}", path.display());
    Ok(())
}

/// 換班敘事（`arriving == true` 為上班）；無職稱或檔未載入時回空字串。
#[must_use]
pub fn get_shift_flavor(title: &str, npc_name: &str, arriving: bool) -> String {
    let g = cache().read().expect("behaviors poisoned");
    let Some(ref bd) = *g else {
        return String::new();
    };
    let Some(role) = bd.roles.get(title) else {
        return String::new();
    };
    let s = if arriving {
        &role.shift_arrive
    } else {
        &role.shift_leave
    };
    s.replace("{name}", npc_name)
}

/// 依職稱讀取 `movement.speed`（對齊既有 `GetMovementDefForTitle` 之 speed）；無則為 1。
#[must_use]
pub fn movement_speed_for_title(title: &str) -> i32 {
    let g = cache().read().expect("behaviors poisoned");
    let Some(ref bd) = *g else {
        return 1;
    };
    let Some(role) = bd.roles.get(title) else {
        return 1;
    };
    role.movement
        .as_ref()
        .map(|m| m.speed)
        .filter(|&s| s > 0)
        .unwrap_or(1)
}

/// 遊戲小時（0–23）映射為 `morning`／`noon` 等（對齊既有 `GetTimePeriod`）。
#[must_use]
pub fn get_time_period(game_hour: i32) -> String {
    let g = cache().read().expect("behaviors poisoned");
    let Some(ref bd) = *g else {
        return "night".to_string();
    };
    for (name, bounds) in &bd.time_periods {
        let start = bounds[0];
        let end = bounds[1];
        if start <= end {
            if game_hour >= start && game_hour < end {
                return name.clone();
            }
        } else if game_hour >= start || game_hour < end {
            return name.clone();
        }
    }
    "night".to_string()
}

fn pick_random_line(lines: &[String]) -> Option<&str> {
    if lines.is_empty() {
        return None;
    }
    let mut rng = rand::rng();
    Some(lines[rng.random_range(0..lines.len())].as_str())
}

/// NPC 進房反應（對齊既有 `PickEnterReaction`）。
#[must_use]
pub fn pick_enter_reaction(title: &str, npc_name: &str) -> String {
    let g = cache().read().expect("behaviors poisoned");
    let Some(ref bd) = *g else {
        return String::new();
    };
    let Some(role) = bd.roles.get(title) else {
        return String::new();
    };
    let Some(line) = pick_random_line(&role.enter_reactions) else {
        return String::new();
    };
    line.replace("{name}", npc_name)
}

/// 閒置動作一句；`disposition` 極端時改讀 morning／night 池（對齊既有 `PickIdleEmote`）。
#[must_use]
pub fn pick_idle_emote(title: &str, period: &str, npc_name: &str, disposition: i32) -> String {
    let g = cache().read().expect("behaviors poisoned");
    let Some(ref bd) = *g else {
        return String::new();
    };
    let Some(role) = bd.roles.get(title) else {
        return String::new();
    };
    let mut effective = period.to_string();
    if disposition < -30 {
        effective = "night".to_string();
    } else if disposition > 30 {
        effective = "morning".to_string();
    }
    let mut pick_from: &[String] = role
        .idle
        .get(effective.as_str())
        .map(Vec::as_slice)
        .unwrap_or(&[]);
    if pick_from.is_empty() {
        pick_from = role.idle.get(period).map(Vec::as_slice).unwrap_or(&[]);
    }
    let Some(line) = pick_random_line(pick_from) else {
        return String::new();
    };
    line.replace("{name}", npc_name)
}

/// 職稱對應的巡邏房間 id 列表（對齊既有 `GetWanderRooms`）。
#[must_use]
pub fn get_wander_rooms(title: &str) -> Vec<String> {
    let g = cache().read().expect("behaviors poisoned");
    let Some(ref bd) = *g else {
        return Vec::new();
    };
    bd.roles
        .get(title)
        .map(|r| r.wander_rooms.clone())
        .unwrap_or_default()
}

/// 巡邏離開／抵達敘事（對齊既有 `GetWanderFlavor`）。
#[must_use]
pub fn get_wander_flavor(title: &str, npc_name: &str, room_name: &str, leaving: bool) -> String {
    let g = cache().read().expect("behaviors poisoned");
    let Some(ref bd) = *g else {
        return String::new();
    };
    let Some(role) = bd.roles.get(title) else {
        return String::new();
    };
    let s = if leaving {
        &role.wander_leave
    } else {
        &role.wander_arrive
    };
    let mut out = s.replace("{name}", npc_name);
    if leaving {
        out = out.replace("{dest}", room_name);
    } else {
        out = out.replace("{from}", room_name);
    }
    out
}
